### PR TITLE
[WIP][MLIR][DISC] Add device type info to memref's address space in lmhlo dialect

### PR DESCRIPTION
**Pull request by azazhu**
We are porting our MLIR-based dynamic shape compiler to tf community (From OP def, Patttern, to Optimization pass, etc).

There will be 3 PRs to process memory allocation for dynamic shape. 1) Mark shape calculation Ops by Op attribute and insert D2H and H2D Op in mhlo dialect; 2) Add device type info to memref's address space in lmhlo dialect; 3) Convert for device alloc/dealloc with memref's address space.

This PR mainly is to add device type info to memref's address space in lmhlo dialect.

Related discussion is [here](https://llvm.discourse.group/t/placement-representation-in-tensor-world-buffer-world/3502/6).

### CODE REVIEW

**Review by joker-eph — COMMENTED** [review]
Some early comment, I haven't read it all but skimming through it feels like the code could get a bit more work. I'll likely get to it after the weekend and be more specific, but feel free to update in the meantime.

**Comment by gbaned** [comment]
@azazhu  Can you please check @joker-eph's comments and resolve conflicts?. Thanks!

**Comment by azazhu** [comment]
> @azazhu Can you please check @joker-eph's comments and resolve conflicts?. Thanks!

Sure, I am working on this PR according to Mehdi's comments.

File: tensorflow/compiler/mlir/hlo/lib/utils/lhlo_utils.cc [hunk 1]
```diff
@@ -0,0 +1,167 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (sanjoy) [inline tensorflow/compiler/mlir/hlo/lib/utils/lhlo_utils.cc:1]: 2021

**Comment by gbaned** [comment]
@azazhu  Can you please check @sanjoy's comments and resolve conflicts?. Thanks!

**Comment by azazhu** [comment]
Sorry for late response, We are working on internal porting(to make sure final porting can e2e run) based on tot TF, which takes more time than we expected. We will continue updating this PR once internal verification finished.

**Comment by gbaned** [comment]
@azazhu  Any update on this PR? Please. Thanks!

**Pull request closed without merging**